### ISSUE TITLE
Question about efficient implementation of spatial Markovian conditioning

### ISSUE THREAD

**Issue by g1239** [LabShuHangGU/MVAR#1]
Great work! Really looking forward to the official code release.

I’m particularly curious about the implementation details of the scale/spatial Markovian conditioning. Specifically:

How did you mask out unnecessary attention computations and still keep different spatial Markovian conditioning attentions running in parallel during inference? Was this done via custom Triton/CUDA kernels, or did you manage it directly within PyTorch (e.g., using selective attention masking or dynamic indexing)?

Would appreciate any details on the technical stack or approach you used. Thanks!

**Comment by NuanBaobao** [comment]
Thanks for your interest in our work! Below are the technical details addressing your questions:

### Scale Markovian Conditioning Implementation
Our implementation builds upon the KV cache mechanism from VAR's baseline. Specifically, we **disable historical scale caching** by maintaining only the current scale's key-value pairs. This is achieved by modifying the standard KV caching logic in [`VAR/models/basic_var.py`](https://github.com/FoundationVision/VAR/blob/df682ce1ed29a8041b1470d71ec75bdfe6921e9d/models/basic_var.py#L107):

```python
if self.caching:
    if self.cached_k is None: self.cached_k = k; self.cached_v = v
    else: k = self.cached_k = torch.cat((self.cached_k, k), dim=dim_cat); v = self.cached_v = torch.cat((self.cached_v, v), dim=dim_cat)
```

### Spatial Markovian Conditioning Implementation

We implement this through two alternative approaches:

1. PyTorch FlexAttention Integration，[FlexAttention](https://pytorch.org/blog/flexattention/)
2. Custom CUDA Kernel Implementation based on existing libraries

**Comment by g1239** [comment]
Thanks for the quick and detailed response — and I really appreciate the clarity and professionalism.

I’ve previously experimented with spatial masking strategies similar to yours using FlexAttention. While the masks did reduce the attention computation, I found that at both 256×256 and 512×512 resolutions, the **inference** runtime overhead introduced by the masking logic actually outweighed the gains, leading to no improvement (and sometimes even a slowdown) in inference speed.

Looking forward to learning a more practical and efficient implementation approach from your project!
